addons/SpellBook/SpellBook.lua

```diff
@@ -0,0 +1,382 @@
+    windower.add_to_chat(7, '//spbk <category> [all] -- Show learnable spells by category. Limited to spells which are learnable, unless all is added after the category.')
+    windower.add_to_chat(7, 'Categories: whitemagic, blackmagic, songs, ninjustu, summoning, bluemagic, geomancy, trusts, all (Trusts are not included in all)')
+    windower.add_to_chat(7, 'Note: A limit on spent job points can be specified for spells learned from Gifts by entering a level of 100-1500. Spells are never given as Gifts for less than 100 jp.')
+end
+
+--[[
+Returns true if the player has any jobs which is high enough level to learn
+the given spell.
+--]]
+function is_learnable(spell)
+    local player_levels = windower.ffxi.get_player().jobs
+    for job,level in pairs(spell.levels) do
+        if player_levels[res.jobs[job].ens] >= level then
+            return true
+        end
+    end
+    return false
+end
+
+--[[
+Formats a spell as the spell's name followed by a list of jobs and levels
+which would qualify to learn that spell.
+--]]
+function format_spell(spell)
+    local format
+
+    if spell.type ~= 'Trust' then
+        local jobs = T{}
+        local levels = T{}
+        for job_id,_ in pairs(spell.levels) do
+            jobs:append(job_id)
+        end
+        jobs:sort()
+        for _,job_id in ipairs(jobs) do
+            if spell.levels[job_id] <= 99 then
+                levels:append(res.jobs[job_id].ens .. ' Lv.' ..
+                    tostring(spell.levels[job_id]))
+            else
+                levels:append(res.jobs[job_id].ens .. ' Jp.' ..
+                    tostring(spell.levels[job_id]))
+            end
+        end
+        format = levels:concat(' / ')
+    else
+        format = ' ( Trust )'
+    end
+    return string.format('%-20s %s', spell.en, format)
+end
+
+--[[
+Show missing spells of a given type. If learnable is true, then the
+results will be limited to spells for which the player has a job at a
+level required to learn the spell.
+--]]
+function spells_by_type(spell_type, learnable_only)
+    local missing_spells = T{}
+    local player_spells = windower.ffxi.get_spells()
+    local spell_count = 0
+
+    for spell_id,spell in pairs(res.spells) do
+        if ((spell_type.type == 'all' and spell.type ~= 'Trust') or
+            spell.type == spell_type.type) and next(spell.levels) ~= nill and
+            not player_spells[spell_id] and (is_learnable(spell) or
+            not learnable_only) and not spell.unlearnable then
+
+            missing_spells:append(format_spell(spell))
+            spell_count = spell_count + 1
+        end
+    end
+
+    if learnable_only then
+        windower.add_to_chat(7, string.format('Showing learnable %s.',
+            spell_type.readable))
+    else
+        windower.add_to_chat(7, string.format('Showing all missing %s.',
+            spell_type.readable))
+    end
+
+    if next(missing_spells) ~= nil then
```

Comment:
You can use `if missing_spells:empty() then`, it evaluates to the same thing but looks more intuitive. It only works on `S{}`, `T{}` and `L{}`, not on regular Lua tables (`{}`). For those you can use `if table.empty({}) then`. You could use this above instead of `next(spell.levels) ~= nil`.